File: impl/matlab/blit.m
% blit gets the next tick of a Hammerich-based BLIT.
% x is the sample number
% sampleRate is self explanatory
% so is frequency
% filterSlope controls how fast the filter falls off, larger values falling
% slower (0.05 is fast, 1.0 is slow)
% harmonicsCount defines the cutoff frequency – it's the number of
% harmonics we want to allow in the final signal
% overlapCount is a measure of resolution, of sorts - how many pulse
% functions do we wantto calculate at once? a higher number gives greater
% accuracy at the expense of computational power
% windowType defines the window type of each individual pulse. currently
% locked to Hamming
function y = blit(x, sampleRate, frequency, filterSlope, harmonicsCount, overlapCount) % windowType
    % find the number of samples in the inter-pulse distance
    s = sampleRate / frequency;
    t = overlapCount * s;
    % make sure x is in the range 1:IPD; we don't need to calculate more
    % than that
    x = mod(x, s);
    % generate our array of sample pointers, one for each overlapping pulse
    ptrs = (1:overlapCount) - round((overlapCount + 1) / 2);
    ptrs = ptrs(:);
    samplePointers = x + (ptrs * s);
    % make sure they're in the right range
    samplePointers(samplePointers > t/2) = samplePointers(samplePointers > t/2) - t;
    % get the value of the pulse at that value
    pulses = hammerichPulse(samplePointers, sampleRate, frequency, filterSlope, harmonicsCount);
    % apply the windowing function to it (currently hardcoded Hamming)
    pulses = pulses .* hamming(samplePointers, overlapCount * s - 1);
    y = sum(pulses, 1);
    % scale so highest value is 1
    largest = max(y);
    y = y .* 1/largest;
end

File: impl/cpp/src/blit.cpp
#include <cmath>
#include <numeric>
#include <vector>

#include "include/blit.hpp"
#include "include/hammerich.hpp"

namespace sub_synth
{
  class blit
  {
    blit(double sample_rate, double frequency, double filter_slope, double harmonics_count, int overlap_count, pulse &pulse)
      : sample_rate(sample_rate)
        , frequency(frequency)
        , filter_slope(filter_slope)
    {}
    
    double get_sample(double sample)
    {
      double s = sample_rate / frequency;
      double x = sample % s;
      
      std::vector<double> sample_pointers;
      for (int i = 1; i <= overlap_count; i++)
      {
        int ptr = i - round((overlap_count + 1) / 2);
        auto val = sample + ptr * s;
        if (val > overlap_count * s / 2)
        {
          val -= overlap_count * s;
        }
        sample_pointers.push_back(val);
      }

      double blit_val = std::accumulate(sample_pointers.begin(), sample_pointers.end(), 0.0, [](double current, double ptr){return current + blit::pulse->get_sample(ptr); /* multiply by window */});
      return blit_val;
    }
  };
} // end namespace sub_synth
